pkg/component/autoscaling/gardenercustommetrics/component.go

```diff
@@ -0,0 +1,215 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gardenercustommetrics implements the gardener-custom-metrics seed component (aka GCMx).
+// For details, see the GardenerCustomMetrics type.
+package gardenercustommetrics
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/component/autoscaling/gardenercustommetrics/kubeobjects"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+	secretsutils "github.com/gardener/gardener/pkg/utils/secrets"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+)
+
+// ComponentName is the component name.
+const ComponentName = componentBaseName
+
+// GardenerCustomMetrics manages an instance of the gardener-custom-metrics component (aka GCMx). The component is
+// deployed on a seed, scrapes the metrics from all shoot kube-apiserver pods, and provides custom metrics by
+// registering as APIService at the custom metrics extension point of the seed kube-apiserver.
+// For information about individual fields, see the NewGardenerCustomMetrics function.
+type GardenerCustomMetrics struct {
+	namespaceName      string
+	containerImageName string
+	kubernetesVersion  *semver.Version
+
+	seedClient     client.Client
+	secretsManager secretsmanager.Interface
+}
+
+// NewGardenerCustomMetrics creates a new GardenerCustomMetrics instance tied to a specific server connection.
+//
+// namespace is where the server-side artefacts (e.g. pods) will be deployed (usually the 'garden' namespace).
+// containerImageName points to the binary for the gardener-custom-metrics pods. The exact version to be used, is
+// determined by contextual configuration, e.g. image vector overrides.
+// If enabled is true, this instance strives to bring the component to an installed, working state. If enabled is
+// false, this instance strives to uninstall the component.
+// seedClient represents the connection to the seed API server.
+// secretsManager is used to interact with secrets on the seed.
+func NewGardenerCustomMetrics(
+	namespace string,
+	containerImageName string,
+	kubernetesVersion *semver.Version,
+	seedClient client.Client,
+	secretsManager secretsmanager.Interface) *GardenerCustomMetrics {
+	return &GardenerCustomMetrics{
+		namespaceName:      namespace,
+		containerImageName: containerImageName,
+		kubernetesVersion:  kubernetesVersion,
+		seedClient:         seedClient,
+		secretsManager:     secretsManager,
+	}
+}
+
+// Deploy implements [component.Deployer.Deploy]()
+func (gcmx *GardenerCustomMetrics) Deploy(ctx context.Context) error {
```

Comment:
Personally, I prefer - single letter names are harder to search for in the IDE. Besides, I'm consistently using GCMx/gcmx as the abbreviation for gardener-custom-metrics. Taking the above into account, are you ok with 'gcmx' here?

Reply:
I don't care too much for now. I just don't know whether it will always still like this as the file evolves and when multiple people start touching it. Typically, we only use one letter, so someone might think about renaming it some day. Anyways, let's keep it as is for now then.